alibi_detect/utils/keops/tests/test_kernels_keops.py

```diff
@@ -64,3 +63,59 @@ def test_gaussian_kernel(gaussian_kernel_params):
             k_xx_argmax, k_xx_min, k_xy_min = k_xx_argmax[0], k_xx_min[0], k_xy_min[0]
         assert (torch.arange(n_instances[0]) == k_xx_argmax.cpu().view(-1)).all()
         assert (k_xx_min >= 0.).all() and (k_xy_min >= 0.).all()
+
+
+if has_keops:
+    class MyKernel(nn.Module):
+        def __init__(self):
+            super().__init__()
+
+        def forward(self, x: LazyTensor, y: LazyTensor) -> LazyTensor:
+            return (- ((x - y) ** 2).sum(-1)).exp()
+
+
+n_features = [5]
+n_instances = [(100, 100), (100, 75)]
+kernel_a = ['GaussianRBF', 'MyKernel']
+kernel_b = ['GaussianRBF', 'MyKernel', None]
+eps = [0.5, 'trainable']
+tests_dk = list(product(n_features, n_instances, kernel_a, kernel_b, eps))
+n_tests_dk = len(tests_dk)
+
+
+@pytest.fixture
+def deep_kernel_params(request):
+    return tests_dk[request.param]
+
+
+@pytest.mark.skipif(not has_keops, reason='Skipping since pykeops is not installed.')
+@pytest.mark.parametrize('deep_kernel_params', list(range(n_tests_dk)), indirect=True)
+def test_deep_kernel(deep_kernel_params):
```

Comment:
Not a huge fan of using this pattern to parametrizing tests, is there a reason not to parametrize each parameter directly? @ascillitoe

Reply:
Yeh the more conventional way to do it would be to parametrize each parameter separately. e.g. see `test_save_model` in `test_saving.py`:

```python
@parametrize('model', [encoder_model])
@parametrize('layer', [None, -1])
def test_save_model(data, model, layer, backend, tmp_path):
```

This approach has the advantage of giving much more descriptive test names which is useful when things go wrong.

We keep writing tests with the current pattern for consistency with existing tests. But, unless we are going to refactor existing tests very soon maybe we should prioritise adopting a new pattern so that we have less refactoring to do later...

Reply:
Ok I was not aware here tbh and followed existing patterns.

Reply:
#610